ert_shared/port_handler.py

```diff
@@ -0,0 +1,88 @@
+import sys
+import socket
+from enum import Enum
+from typing import Optional
+
+
+class Service(Enum):
+    storage = 0
+    evaluator = 1
+    dask_scheduler = 2
+
+
+class PortHandler:
+    def __init__(self):
+        self._range = range(51820, 51840 + 1)
+        self._used = {}
+
+    def assign_port(self, service: Service, host: str = None, port: int = None) -> int:
+
+        if service in self._used:
+            return self._used[service]["port"]
+
+        if host is None:
+            host = PortHandler._get_ip_address()
+
+        if port is not None:
+            # Port provided - forcing which port to assign. Test purposes only!
+            self._used[service] = {"port": port, "host": host}
+            return port
+
+        for port in self._range:
+            if port not in self._used.values():
+                try:
+                    sock = PortHandler._bind_socket(port=port, host=host)
+                    self._used[service] = {"port": port, "host": host}
+                    return port
+                except OSError:
+                    continue
+        sys.exit(f"No ports available in {self._range}. Quitting.")
```

Comment:
should we raise a `regular` custom exception and let the consumer of port_handler decide what to do?

Reply:
agree - will fix as soon as i manage to get all tests existing green - then i will write tests targeting the port_handler and fix this 👍

Reply:
Maybe `IndexError` would fit here?